README.md
# React API Manager

APIManager used in most React.js projects. Basically a `fetch` variant that's able to handle retries and timeouts. [CHANGELOG.md](https://github.com/skizzo/react-api-manager/blob/master/CHANGELOG.md "react-api-manager Change Log")

## Installation

`npm install --save react-api-manager`


## Usage

### APIManager.init (options) : void

#### Parameter options

- `apiroot`: root to your API, e.g. `https://nerdybirdy.com/api`

- `debug`: debug mode (default: `false`)


### APIManager.fetchAPI (params) : Promise

#### Parameter params

- `path`: API path **without leading slash**, e.g `projects/my-test-project`

- `params`: (Optional) array of parameters

- `retries`: (Optional) Amount of retries in case fetch fails (default: `5`)

- `timeout`: (Optional) Amount of milli seconds until fetch fails with Timeout error (default: `10000` aka 10s)

#### Returns

Returns a promise that either 

- **resolves** and contains data in JSON format, or

- **rejects** and contains error data, possible values for **`error.message`**: 

  - `Timeout`: All retries have been tried, request still timed out.

  - `Server` (500): There was an error on the server, in this case the error object also contains the properties `file`, `line` and `message`.
  Check this out, next line.

  - `MethodNotFound` (404): The API method could not be found on server.

  - `Unknown`: Some weird shit happened.

## Contributing

This package is customized for a SLIM REST API and should not be contributed to.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [v0.0.8]

- Removed log-with-style, otherwise jest tests fail because there is no "window" :/

## [v0.0.3]

- Check check

## [v0.0.2]

- Added dependencies (react, whatwg-fetch, log-with-style)
- Added APIManager.js itself
- Currently only works for GET requests

## [v0.0.1]

- Added index.js


__tests__/index.test.js
'use strict';

var {APIManager} = require ("../index")
const VALID_API_URL = "https://nerdybirdy.net/api/"

describe ("APIManager", () => {

	it ("itself", () => {
		expect (APIManager).toBeDefined ()
		expect (APIManager.init).toBeDefined ()
		// expect (1).toBe (1)
	})

	it ("init()", () => {
		expect (APIManager.init).toBeDefined ()
	})

	it ("init() rejected when no params given", (done) => {
		APIManager.init ()
		.then (res => {
			console.log ("res: " + JSON.stringify (res, null, 2))
		})
		.catch (err => { // CustomError
			expect (err).toBeDefined ()
			expect (err.message).toEqual ("NoApiRoot")
			// console.error ("err: ", err)
			done()
		})
	})

	it ("init() resolves when params given", (done) => {
		APIManager.init ({apiroot: VALID_API_URL})
		.then (() => {
			done ()
		}) 
		.catch (err => { // CustomError
			console.error (err)
		})
	}, 10000) // can take a while.

	it ("fetchAPI() with bullshit params throws error", (done) => {
		APIManager.init ({apiroot: "https://does-not-exist/"})
		.then (() => {}) 
		.catch (err => { // CustomError
			expect (err).toBeDefined ()
			expect (err.message).toEqual ("NetworkRequestFailed")
			done ()
		})
	}, 10000) // can take a while.

	it ("fetchAPI() with tiny timeout and no retries fails", (done) => {
		APIManager.init ({apiroot: VALID_API_URL}).then (() => {
			APIManager.fetchAPI ({path: "status", timeout: 10, retries: 0})
			.then (res => {})
			.catch (err => {done ()})
		}) 
		.catch (err => { // CustomError
		})
	}, 10000) // can take a while.

	//	from now on, APIManager is initialized with valid API path
	it ('handles server error (500)', (done) => { 
		APIManager.fetchAPI ({path: "mock/servererror"})
		.then (resJSON => {
		})
	  .catch (error => {
			expect (error).toBeDefined()
			expect (error.message).toEqual ("Server")
			expect (error.data).toBeDefined ()
			expect (error.data.message).toEqual ("Use of undefined constant dummy - assumed 'dummy'")
			expect (error.data.status).toEqual (500)
	  	done ()
	  }) 
	}, 5000)

	it ('handles non-existing server functions (404)', (done) => { 
		APIManager.fetchAPI ({path: "foo/bar"})
		.then (resJSON => {
			expect (resJSON).toBeUndefined ()
		})
	  .catch (error => { 
			expect (error).toBeDefined()
			expect (error.message).toEqual ("MethodNotFound")
			expect (error.data).toBeDefined ()
			expect (error.data.status).toEqual (404)
	  	done ()
	  })
	}, 5000)


})

src/APIManager.js
const VERSION = "0.0.8"

require ("whatwg-fetch")

class CustomError extends Error {
  constructor (data = null, ...args) {
    super (...args);
    this.data = data;
  }
}

const getTimeoutPromise = (ms, promise, contentObj = {type: "Timeout"}) => {
  return new Promise ((resolve, reject) => {
    setTimeout (() => {
      reject (new CustomError (null, "Timeout"))
    }, ms)
    promise.then (resolve, reject)
  })
}
/*
const getFetchPostParams = (params = {}) => {
  return {
    method: 'POST',
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      // 'Upgrade-Insecure-Requests': '1',
    },
    body: JSON.stringify (params)
  }
}
*/

var APIManager = {

  init (options = {}) {
    const {apiroot, debug} = options

    if (!apiroot) {
      return new Promise ((resolve, reject) => {
        reject (new CustomError (null, "NoApiRoot"))
      })
    }

    this.apiroot = apiroot
    this.debug = !!debug ? debug : false

    this.debugger ("init ()", Object.assign ({}, options, {version: VERSION}))
    return this.fetchAPI ({path: "status"})
  },

  debugger (msg, data = null, type = "default") { // 
    if (!this.debug)
      return

    const colorBG = type == "error" ? "red" : "black"
    const colorFG = "white"

    console.log (`[APIManager.js]: ` + msg, data)
  },

  fetchAPI (params) { // ONLY WORKS FOR GET REQUESTS!

    if (!this.apiroot) {
      return new Promise ((resolve, reject) => {
        this.debugger ("fetchAPI (): missing apiroot param in init ()", params, "error")
        reject (new CustomError (null, "NoApiRoot"))
      })
    }

    const fetchURLAdd = !params.params ? "" :"?" + Object.keys (params.params).map (k => encodeURIComponent (k) + "=" + encodeURIComponent (params.params[k])).join ("&")

    if (!params.path)                 params.path = ""
    if (!params.params)               params.params = {}
    if (!params.timeout)              params.timeout = 10000
    if (params.retries == undefined)  params.retries = 5

    params.pathFull = `${this.apiroot}${params.path}${fetchURLAdd}`

    return getTimeoutPromise (params.timeout, fetch (params.pathFull))
    .then (res => { 
      const {status} = res
      switch (status) {
        case 200: { return res.json(); break }
        case 500: { throw new CustomError ({res, status}, "Server"); break }
        case 404: { throw new CustomError ({res, status}, "MethodNotFound"); break }
        default:  { throw new CustomError ({res, status}, "Unknown") } 
      }
    })
    .then (resJSON => {
      return resJSON 
    })
    .catch ((error) => {
      const {data, message} = error 
      if (message == "Timeout") {
        if (params.retries > 0) {
          return this.fetchAPI (Object.assign ({}, params, {retries: params.retries - 1, timeout: params.timeout + 1000}) )
        }
        else {
          throw error
        }
      }
      else if (message == "Server") { // thrown in getTimeoutPromise()'s first .then(..)
        return data.res.json()
        .then ((errorJSON) => {
          const {response} = errorJSON
          const {errorFull} = response || {errorFull: {}}
          const {file, line, message} = errorFull
          // throw new CustomError ({...data, file, line, message}, "Server")
          throw new CustomError (Object.assign ({}, data, {file, line, message}), "Server")
        })
      }
      else if (!data) {
        throw new CustomError ({}, "NetworkRequestFailed")
      }
      else {
        throw error
      }
    })
  },
}
module.exports = {
  APIManager,
}